sdks/go/pkg/beam/core/runtime/genx/genx.go

```diff
@@ -0,0 +1,172 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package genx is a convenience package to better support the code
+// generator. It can be depended on by the user facing beam package
+// and be refered to by generated code.
+//
+// Similarly, it can depend on beam internals and access the canonical
+// method list in the graph package, or other packages to filter out
+// types that aren't necessary for registration (like context.Context).
+package genx
+
+import (
+	"reflect"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/funcx"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/graph"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime"
+)
+
+// RegisterDoFn is a convenience function for registering DoFns.
+// Differs from RegisterFunction and RegisterType by introspecting
+// all parameters and returns of Lifecycle methods on the dofn,
+// and registers those types for you.
+//
+// Panics if not passed a dofn.
+func RegisterDoFn(dofn interface{}) {
+	f, ts, err := registerDoFn(dofn)
+	if err != nil {
+		panic(err)
+	}
+	if f != nil {
+		runtime.RegisterFunction(f)
+	}
+	for _, t := range ts {
+		runtime.RegisterType(t)
+	}
+}
+
+// registerDoFn returns all types associated with the provided DoFn.
+// If passed a functional DoFn, the first return is a Function to
+// register with runtime.RegisterFunction.
+// The second return is all types to register with runtime.RegisterType.
+// Returns an error if the passed in values are not DoFns.
+func registerDoFn(dofn interface{}) (interface{}, []reflect.Type, error) {
+	if rt, ok := dofn.(reflect.Type); ok {
+		if rt.Kind() == reflect.Ptr {
+			rt = rt.Elem()
+		}
+		dofn = reflect.New(rt).Interface()
+	}
+	fn, err := graph.NewFn(dofn)
+	if err != nil {
+		return nil, nil, err
+	}
+	c := cache{}
+	var valid bool
+	// Validates that this is a DoFn or combineFn.
+	do, err := graph.AsDoFn(fn, graph.MainUnknown)
+	if err == nil {
+		valid = true
+		handleDoFn(do, c)
+	} else if cmb, err2 := graph.AsCombineFn(fn); err2 == nil {
+		valid = true
+		handleCombineFn(cmb, c)
+	}
+	if !valid {
+		// Return the DoFn specific error as that's more common.
+		return nil, nil, err
+	}
+
+	var retFunc interface{}
+	rt := reflect.TypeOf(dofn)
+	switch rt.Kind() {
+	case reflect.Func:
+		retFunc = dofn
+		c.regFuncTypes(rt)
+	default:
+		c.regType(rt)
+	}
+	var retTypes []reflect.Type
+	for _, t := range c {
+		retTypes = append(retTypes, t)
+	}
+	return retFunc, retTypes, nil
+}
+
+func handleDoFn(fn *graph.DoFn, c cache) {
+	c.pullMethod(fn.SetupFn())
+	c.pullMethod(fn.StartBundleFn())
+	c.pullMethod(fn.ProcessElementFn())
+	c.pullMethod(fn.FinishBundleFn())
+	c.pullMethod(fn.TeardownFn())
+	if !fn.IsSplittable() {
+		return
+	}
+	sdf := (*graph.SplittableDoFn)(fn)
+	c.pullMethod(sdf.CreateInitialRestrictionFn())
+	c.pullMethod(sdf.CreateTrackerFn())
```

Comment:
RTs would be registered anyway since they show up in the ProcessElement signature. It's OK for extra types to be registered, the framework just won't automatically synthesize them like it needs to for other elements.